src/main/java/com/mraof/minestuck/tileentity/redstone/AreaEffectTileEntity.java

```diff
@@ -0,0 +1,195 @@
+package com.mraof.minestuck.tileentity.redstone;
+
+import com.mraof.minestuck.block.redstone.AreaEffectBlock;
+import com.mraof.minestuck.effects.CreativeShockEffect;
+import com.mraof.minestuck.effects.MSEffects;
+import com.mraof.minestuck.tileentity.MSTileEntityTypes;
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.network.NetworkManager;
+import net.minecraft.network.play.server.SUpdateTileEntityPacket;
+import net.minecraft.potion.Effect;
+import net.minecraft.potion.EffectInstance;
+import net.minecraft.tileentity.ITickableTileEntity;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+
+public class AreaEffectTileEntity extends TileEntity implements ITickableTileEntity
+{
+	//TODO deserialize/reserialize function + triggering of all summoners
+	
+	private Effect effect;
+	private int effectAmplifier;
+	private BlockPos minEffectPos;
+	private BlockPos maxEffectPos;
+	
+	public AreaEffectTileEntity()
+	{
+		super(MSTileEntityTypes.AREA_EFFECT.get());
+	}
+	
+	@Override
+	public void tick()
+	{
+		if(level == null || !level.isAreaLoaded(getBlockPos(), 0))
+			return;
+		
+		if(level.getGameTime() % 80 == 0 && !level.isClientSide && level.getBlockState(getBlockPos()).getBlock() instanceof AreaEffectBlock && getBlockState().getValue(AreaEffectBlock.POWERED))
+		{
+			giveEntitiesEffect();
+		}
+	}
+	
+	public void giveEntitiesEffect()
```

Comment:
Effect-adding code here is pretty much being duplicated in here. You should be able to write it better with an appropriately extracted function.